### PR TITLE
MHA forward pass bug fix

### PR DESCRIPTION
**Pull request by george-qi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77761

### CODE REVIEW

**Review by jbschlosser — COMMENTED** [review]
Thanks for the fix!

File: torch/nn/modules/activation.py [hunk 1]
```diff
@@ -1066,9 +1066,9 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
             # enforce that the dtypes all match or test cases where
             # they don't!
             why_not_fast_path = "non-self attention was used (query, key, and value are not the same Tensor)"
-        elif query.dtype != self.in_proj_bias.dtype:
+        elif self.in_proj_bias is not None and query.dtype != self.in_proj_bias.dtype:
             why_not_fast_path = f"dtypes of query ({query.dtype}) and self.in_proj_bias ({self.in_proj_bias.dtype}) don't match"
-        elif query.dtype != self.in_proj_weight.dtype:
+        elif self.in_proj_bias is not None and query.dtype != self.in_proj_weight.dtype:
```
[the remark this answers is not in the document]
Comment (george-qi) [reply torch/nn/modules/activation.py]: You got to it before I finished all my fixes :) Thanks, this is fixed!

**Comment by george-qi** [comment]
> Looks like some `third_party` changes made it in - these should be removed :)
> 
> Also do you mind adding a test or two for the None cases?

Fixed the `third_party` changes and added a test for both cases :)

File: test/test_nn.py [hunk 2]
```diff
@@ -20537,6 +20537,21 @@ def test_multihead_attn_fast_path_query_and_bias_have_different_dtypes(self, dev
         query = torch.randn(3, 3, 3, dtype=dtype, device=device)
         mha(query, query, query)
 
+    @dtypes(torch.double)
+    @torch.no_grad()
+    def test_multihead_attn_in_proj_bias_none(self, device, dtype):
+        mha = torch.nn.MultiheadAttention(1, 1, bias=True, dtype=dtype, device=device)
```
Comment (jbschlosser) [inline test/test_nn.py:20543]: I think you want `bias=False` here to test this case, no?

File: test/test_nn.py [hunk 3]
```diff
@@ -20537,6 +20537,21 @@ def test_multihead_attn_fast_path_query_and_bias_have_different_dtypes(self, dev
...
+        query = torch.rand(3, 2, 1)
+        mha(query, query, query)
+
+    @dtypes(torch.double)
+    @torch.no_grad()
+    def test_multihead_attn_in_proj_weight_none(self, device, dtype):
+        mha = torch.nn.MultiheadAttention(4, 4, vdim=2, kdim=2, dtype=dtype, device=device)
```
Comment (jbschlosser) [inline test/test_nn.py:20550]: The MHA code in general is pretty hard to read :)

For clarity, I think it'd be useful to have a comment here explaining that picking kdim and vdim != embed_dim will cause the logic to use per-input projection weights, and `in_proj_weight` will be None

File: test/test_nn.py (hunk 2 above)
Comment (george-qi) [reply test/test_nn.py]: Yeah sorry, forgot to push that

File: test/test_nn.py (hunk 3 above)
Comment (george-qi) [reply test/test_nn.py:20553]: Done, thanks :)

**Review by jbschlosser — APPROVED** [review]
LGTM

**Pull request closed without merging**